### PR TITLE
eliminate dependency on go-openapi/spec

### LINKED ISSUE #96303: Eliminate dependency on go-openapi/spec
Thanks to the work done in https://github.com/kubernetes/kubernetes/pull/95954 we should be able to completely remove the dependency on `go-openapi/spec`. It requires to bring back a few types (and related constructors) that were removed in https://github.com/kubernetes/kube-openapi/commit/73fa09a8495ec96b58cf5844be1ec467520a0541.

Refer to https://kubernetes.slack.com/archives/C0EG7JC6T/p1604659881491000?thread_ts=1592287631.273100&cid=C0EG7JC6T for additional context.

### PR DESCRIPTION
**Pull request by gautierdelorme**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

We don't need to rely on `github.com/go-openapi/spec` anymore since `k8s.io/kube-openapi` supports all the features required by Kubernetes (https://github.com/kubernetes/kube-openapi/pull/234).

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubernetes/issues/96303

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

Signed-off-by: Gautier Delorme <gautier.delorme@gmail.com>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -512,7 +512,7 @@ replace (
 	k8s.io/klog/v2 => k8s.io/klog/v2 v2.8.0
 	k8s.io/kube-aggregator => ./staging/src/k8s.io/kube-aggregator
 	k8s.io/kube-controller-manager => ./staging/src/k8s.io/kube-controller-manager
-	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20210305001622-591a79e4bda7
+	k8s.io/kube-openapi => github.com/gautierdelorme/kube-openapi v0.0.0-20210418101956-5b3ab66456a2
```
Comment (apelisse) [inline go.mod:515]: I'm sure that's a mistake ;-)

**Review by apelisse — COMMENTED** [review]
cc @sttts

File: go.mod (hunk 1 above)
Comment (gautierdelorme) [reply go.mod:515]: @apelisse this is a WIP PR 😊 Obviously I'll use kubernetes/kube-openapi once the other PR is merged. I linked to the two blockers in the PR description.

**Comment by gautierdelorme** [comment]
/test pull-kubernetes-integration

flake https://github.com/kubernetes/kubernetes/issues/100556

**Comment by gautierdelorme** [comment]
/test pull-kubernetes-e2e-kind

flake https://github.com/kubernetes/kubernetes/issues/101275

**Comment by gautierdelorme** [comment]
/test pull-kubernetes-integration

flake https://github.com/kubernetes/kubernetes/issues/100556

**Comment by dims** [comment]
/assign @sttts @liggitt 

thanks @gautierdelorme assigning to the right folks. looks very mechanical so i am happy to lgtm after approval

File: vendor/modules.txt [hunk 2]
```diff
@@ -419,7 +419,6 @@ github.com/go-openapi/jsonreference
 # github.com/go-openapi/loads => github.com/go-openapi/loads v0.19.4
 # github.com/go-openapi/runtime => github.com/go-openapi/runtime v0.19.4
 # github.com/go-openapi/spec v0.19.5 => github.com/go-openapi/spec v0.19.5
-## explicit
 github.com/go-openapi/spec
```
Comment (liggitt) [inline vendor/modules.txt:422]: we didn't actually drop anything under `vendor/github.com/go-openapi/spec` as part of this, so we actually increased our dependencies... what is requiring `github.com/go-openapi/spec` now?

Comment (gautierdelorme) [reply vendor/modules.txt:422]: @liggitt this is what I shared in Slack:

`github.com/go-openapi/spec` is still being pulled in by `kustomize` though (but k/k doesn't depend on it anymore).

```console
$ go mod why github.com/go-openapi/spec
# github.com/go-openapi/spec
k8s.io/kubernetes/pkg/kubectl/cmd/convert
k8s.io/cli-runtime/pkg/resource
sigs.k8s.io/kustomize/api/krusty
sigs.k8s.io/kustomize/kyaml/openapi
github.com/go-openapi/spec
```

Not sure if that's something we want to change in the future (https://github.com/kubernetes-sigs/kustomize/blob/f93cee94404bf0b60ff82785be9ff2af9797463f/kyaml/openapi/openapi.go#L14).

Comment (liggitt) [reply vendor/modules.txt:422]: ah, right... https://github.com/kubernetes/kubernetes/pull/98946#discussion_r589798392

tracked in https://github.com/kubernetes-sigs/kustomize/issues/3705

**Comment by liggitt** [comment]
generated openapi files still reference this, xref https://github.com/kubernetes/kubernetes/issues/101375

**Pull request merged**